File: PO/Project/C++/src/animal.cpp
#include "animal.h"


animal::animal(int strength, int initiative, YX position, world* world_point)
: organizm(world_point, strength, initiative, position)
{
}


void animal::Draw(YX position)
{   
    if (strength > starting_strength) 
    {
        attron(A_BOLD); 
    }

    attron(COLOR_PAIR(1));
    
    mvaddch(position.y, position.x, avatar);
    
    attroff(COLOR_PAIR(1));

    if (strength > starting_strength) 
    {
        attroff(A_BOLD); 
    }
}

int animal::Move(YX delta) //0 - normal move, 1 - collision
{
    bool collided = false;

    world_point->FindField(position)->member = nullptr;

    YX new_position = {position.y + delta.y, position.x + delta.x};

    if (new_position.x >= world_point->GetDimensions().x)
    {
        new_position.x = world_point->GetDimensions().x - 1;
        //mvprintw(0, 0, "right wall");
    }
    else if (new_position.x < 0)
    {
        new_position.x = 0;
        //mvprintw(0, 0, "left wall");
    }

    if (new_position.y >= world_point->GetDimensions().y)
    {
        new_position.y = world_point->GetDimensions().y - 1;
        //mvprintw(2, 0, "bottom wall");
    }
    else if (new_position.y < 0)
    {
        new_position.y = 0;
        //mvprintw(2, 0, "top wall");
    }

    //fight with the member of the new position (must return 1 if attak is failed and 0 is attak is succesful)
    organizm* target = world_point->FindField(new_position)->member;

    if (target != nullptr)
    {
        collided = true;
    }

    if ((collided) && (Collision(target) == 1))
    {
        // //diagnostics:
        // printw("\nattaker is dead.");
        // getch();

        return 1;
    }
    
    position = new_position;

    world_point->FindField(position)->member = this;

    if (collided)
    {
        return 1;
    }
    return 0;
}

int animal::Action() // 0 - normal move, 1 - collision detected
{
    int delta_y = rand() % 3 - 1; // Generates random number between -1 and 1
    int delta_x = rand() % 3 - 1; // Generates random number between -1 and 1

    if ((delta_x == 0) && (delta_y == 0)) {
        return 0;
    }

    return Move({delta_y, delta_x});
}

int animal::Collision(organizm* target)
{
    world_point->GetLogger()->LogCollision(this, target);
    
    int result = Fight(target);
    
    world_point->GetLogger()->LogCollisionResult((result) ? target : this);

    return result;
}

int animal::Fight(organizm* target)
{
    if (target->RejectAttak(this)) {
        if (death_sentence) {
            world_point->Destroy(this);
            return 1;
        }

        world_point->FindField(position)->member = this;

        return 1;
    }

    if (target == nullptr)
    {
        return 0;
    }

    if (target->GetStrength() <= strength)
    {
        // succes
        //world_point->Kill(target);
        world_point->Destroy(target);
        return 0;
    }
    else if (target->GetStrength() > strength)
    {
        // failed
        //world_point->Kill(this);
        world_point->Destroy(this);
        return 1;
    }
}

CLASS animal::GetClass()
{
    return ANIMAL; 
}

animal::~animal()
{
}

File: PO/Project/JavaProject/src/Animals/Animal.java
package Animals;

import Abstractions.*;

import java.awt.*;
import java.util.concurrent.ThreadLocalRandom;

public class Animal extends Organism {

    public Animal(int strength, int initiative, Point position, World world) {
        super(world, strength, initiative, position, "Anon.png");
    }


    public int move(Point delta) {
        boolean collided = false;
        world.findField(position).setMember(null);

        Point newPosition = new Point(position.x + delta.x, position.y + delta.y);

        if (newPosition.x >= world.getDimensions().width) {
            newPosition.x = world.getDimensions().width - 1;
            delta.x = 0;
            System.out.println("collision Right!");
        } else if (newPosition.x < 0) {
            newPosition.x = 0;
            delta.x = 0;
            System.out.println("collision Left!");
        }

        if (newPosition.y >= world.getDimensions().height) {
            newPosition.y = world.getDimensions().height - 1;
            delta.y = 0;
            System.out.println("collision Bottom!");
        } else if (newPosition.y < 0) {
            newPosition.y = 0;
            delta.y = 0;
            System.out.println("collision Top!");
        }

        if (delta.x == 0 && delta.y == 0) {
            world.findField(position).setMember(this);
            return 0;
        }

        Organism target = world.findField(newPosition).getMember();

        if (target != null) {
            collided = true;
        }

        if (collided && collision(target) == 1) {
            return 1;
        }

        position = newPosition;
        world.findField(position).setMember(this);

        return collided ? 1 : 0;
    }

    @Override
    public int action() {
        int deltaX, deltaY;
        do {
            deltaY = ThreadLocalRandom.current().nextInt(-1, 2); // Generates random number between -1 and 1
            deltaX = ThreadLocalRandom.current().nextInt(-1, 2); // Generates random number between -1 and 1
        } while ((deltaX == 0 && deltaY == 0) || (world.findField(new Point(position.x + deltaX, position.y + deltaY)) == null));

        return move(new Point(deltaX, deltaY));
    }

    @Override
    public int collision(Organism target) {
        world.getLogger().logCollision(this, target);

        if (world.sameType(this, target)) {
            reproduct();
            world.getLogger().logReproductionCollision(this, target);
            return 1;
        }

        int result = fight(target);
        world.getLogger().logCollisionResult((result == 1) ? target : this);

        return result;
    }

    @Override
    public int fight(Organism target) {
        if (target.rejectAttack(this)) {
            if (deathSentence) {
                world.destroy(this);
                return 1;
            }
            world.findField(position).setMember(this);
            return 1;
        }

        if (target == null) {
            return 0;
        }

        if (target.getStrength() <= strength) {
            world.destroy(target);
            return 0;
        } else {
            world.destroy(this);
            return 1;
        }
    }

    @Override
    public boolean rejectAttack(Organism attacker) {
        return false;
    }

    @Override
    public ClassType getClassType() {
        return ClassType.ANIMAL;
    }

    @Override
    public Organism makeChild(Point newPosition) {
        return new Animal(strength, initiative, newPosition, world); // Adjust parameters as necessary
    }
}
